iothub/device/src/BaseClient.cs

```diff
@@ -0,0 +1,1083 @@
+|            |       |             |      |             |       |            |       |              |
+| overrides: |       | overrides:  |      |  overrides  |       | overrides: |       | overrides:   |
+|  Open      |       |  Open       |      |   Open      |       |  Open      |       |  Receive     |
+|  Close     |       |  SendEvent  |      |   SendEvent |       |            |       |              |
+|            |       |  SendEvents |      |   SendEvents|       +------------+       +--^--^---^----+
++------------+       |  Receive    |      |   Receive   |                               |  |   |
+                     |  Reject     |      |   Reject    |                               |  |   |
+                     |  Abandon    |      |   Abandon   |                               |  |   |
+                     |  Complete   |      |   Complete  |                               |  |   |
+                     |             |      |             |                               |  |   |
+                     +-------------+      +-------------+     +-------------+-+inherits-+  |   +---inherits-+-------------+
+                                                              |             |              |                |             |
+                                                              | AMQP        |              inherits         | HTTP        |
+                                                              | Transport   |              |                | Transport   |
+                                                              | Handler     |          +---+---------+      | Handler     |
+                                                              |             |          |             |      |             |
+                                                              | overrides:  |          | MQTT        |      | overrides:  |
+                                                              |  everything |          | Transport   |      |  everything |
+                                                              |             |          | Handler     |      |             |
+                                                              +-------------+          |             |      +-------------+
+                                                                                       | overrides:  |
+                                                                                       |  everything |
+                                                                                       |             |
+                                                                                       +-------------+
+TODO: revisit DefaultDelegatingHandler - it seems redundant as long as we have to many overloads in most of the classes.
+*/
+
+    /// <summary>
+    /// Contains methods that a device can use to send messages to and receive from the service.
+    /// </summary>
+    public class BaseClient : IDisposable
```

Comment:
>public [](start = 4, length = 6)

Should this be internal?  (assuming the language allows a public object to inherit from an internal base).  I'm mostly concerned that making this public means that we need to worry about breaking changes at this level.

Reply:
Yeah, unfortunately a public class cannot derive from an internal class.

Reply:
here's why: [https://ericlippert.com/2012/11/13/why-is-deriving-a-public-class-from-an-internal-class-illegal/](https://ericlippert.com/2012/11/13/why-is-deriving-a-public-class-from-an-internal-class-illegal/)

have you looked at using composition instead? is that too much overhead?

---
In reply to: [185608658](https://github.com/Azure/azure-iot-sdk-csharp/pull/465#discussion_r185608658) [](ancestors = 185608658)

Reply:
Yes, that is the plan.. I mentioned to Tamer and Bert as well. I will send an update with composition by tomorrow.

Reply:
Changed the code to make it internal.